ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/TransactionImple.java

```diff
@@ -1032,6 +1032,40 @@ public String toString()
 		}
 	}
 
+	public void endAssociation(Xid _tranID, XAResource _theXAResource, int xaState, int txInfoState) throws XAException {
+		List<XAResource> xars = new ArrayList<XAResource>();
+		int txInfo = getXAResourceState(_theXAResource);
+		if ((txInfo != TxInfo.NOT_ASSOCIATED) && (txInfo != TxInfo.FAILED)) {
+			xars.add(_theXAResource);
+		}
+		for (Object resource : _duplicateResources.keySet()) {
+			XAResource dupXar = (XAResource) resource;
+			if (_theXAResource.isSameRM(dupXar)) {
+				xars.add(dupXar);
+			}
+		}
+
+		for (XAResource xar : xars) {
+			try {
+				xar.end(_tranID, xaState);
+				setXAResourceState(xar, txInfoState);
+			} catch (XAException e1) {
+				switch (e1.errorCode) {
+					case XAException.XA_RBROLLBACK:
+					case XAException.XA_RBCOMMFAIL:
+					case XAException.XA_RBDEADLOCK:
+					case XAException.XA_RBINTEGRITY:
+					case XAException.XA_RBOTHER:
+					case XAException.XA_RBPROTO:
+					case XAException.XA_RBTIMEOUT:
+					case XAException.XA_RBTRANSIENT:
+						setXAResourceState(xar, txInfoState);
+				}
+				throw e1;
```

Comment:
Why aren't you calling end on the remaining resources in xars if one of them fails to end cleanly?